src/python/pants/backend/jvm/tasks/jvm_compile/zinc/zinc_compile.py

```diff
@@ -415,13 +415,26 @@ def _compile_hermetic(self, jvm_options, ctx, classes_dir, jar_file,
     directory_digests = [
       entry.directory_digest for entry in relevant_classpath_entries if entry.directory_digest
     ]
+
     if len(directory_digests) != len(relevant_classpath_entries):
       for dep in relevant_classpath_entries:
         if not dep.directory_digest:
-          raise AssertionError(
-            "ClasspathEntry {} didn't have a Digest, so won't be present for hermetic "
-            "execution of zinc".format(dep)
-          )
+          if not os.path.exists(dep.path):
+            raise AssertionError(
+              "ClasspathEntry {} didn't have a Digest, so won't be present for hermetic "
+              "execution of zinc".format(dep)
+            )
+          directory_digests.append(dep.directory_digest)
+          relpath = fast_relpath(dep.path, get_buildroot())
+          classes_dir_snapshot, = self.context._scheduler.capture_snapshots([
```

Comment:
This isn't the right place to fix this, unfortunately.

Hermetic compiles already capture their outputs when they run, so re-capturing here would be a bandaid.

I'd suggest looking for calls to "materialize", and ensuring that the things that are getting materialized end up written to the relevant `ClasspathEntrys`.

Reply:
where the `materialize` happens, this is likely correct, but do you have any pointer on when v1 cache was hit, how is `ClasspathEntry` populated?

Reply:
Afaict, `--cache-compile-rsc-ignore` works fine before, meaning whenever materialize is happening, ClasspathEntry is set correctly when the digest. So the issue might be more related to hitting v1 cache.

Reply:
Adjust the change to happen during validation:
https://github.com/pantsbuild/pants/blob/94bc795caf30e2853c399494bd4822832b33a417/src/python/pants/backend/jvm/tasks/jvm_compile/jvm_compile.py#L489